tests/bls_tests.rs

```diff
@@ -0,0 +1,262 @@
+#![cfg(feature = "spec_tests")]
+use ethereum_consensus::crypto::{
+    aggregate, aggregate_verify, fast_aggregate_verify, PublicKey, SecretKey, Signature,
+};
+use glob::glob;
+use hex;
+use serde::{Deserialize, Serialize};
+use serde_yaml;
+use std::fmt::Debug;
+use std::fs::File;
+fn decode_hex_with_prefix<T: AsRef<[u8]> + Debug>(data: T) -> Vec<u8> {
+    let data = data.as_ref();
+    let data = if data.starts_with(b"0x") {
+        &data[2..]
+    } else {
+        data
+    };
+    hex::decode(data).expect("is not well-formed hex")
+}
+#[derive(Debug, Serialize, Deserialize)]
+struct SigningInput {
+    privkey: String,
+    message: String,
+}
+#[derive(Debug, Serialize, Deserialize)]
+struct SigningTestIO {
+    input: SigningInput,
+    output: String,
+}
+impl SigningTestIO {
+    fn verify(&self) {
+        // let input = &self.input;
+        let secret_key_bytes = decode_hex_with_prefix(&self.input.privkey);
+        let secret_key = match SecretKey::from_bytes(&secret_key_bytes) {
+            Ok(sk) => sk,
+            // this is the empty secret key case
+            Err(_) => {
+                assert_eq!(self.output, "~");
+                return;
+            }
+        };
+        let message_bytes = decode_hex_with_prefix(&self.input.message);
+        let signature = secret_key.sign(&message_bytes);
+        let signature_bytes = decode_hex_with_prefix(&self.output);
+        let expected_signature = Signature::from_bytes(&signature_bytes).unwrap();
+        assert_eq!(signature, expected_signature)
+    }
+}
+#[derive(Debug, Serialize, Deserialize)]
+struct AggregatingTestIO {
+    input: Vec<String>,
+    output: String,
+}
+impl AggregatingTestIO {
+    fn verify(&self) {
+        let input_signatures: Vec<Signature> = self
+            .input
+            .iter()
+            .map(|x| Signature::from_bytes(&decode_hex_with_prefix(&x)).unwrap())
+            .collect();
+        let aggregate = match aggregate(&input_signatures) {
+            Ok(agg) => agg,
+            // handling for zero sized input and output
+            Err(_) => {
+                assert_eq!(self.output, "~");
```

Comment:
can you show me which test case has the string `"~"`? i think this is some artifact of the yaml en/decoding

Reply:
It's this: https://media.githubusercontent.com/media/ethereum/consensus-spec-tests/master/tests/general/phase0/bls/aggregate/small/aggregate_na_signatures/data.yaml